### PR TITLE
fix Type3 Beamer glyph rendering

### PR DESCRIPTION
**Pull request by imkerberos**
## Summary

Fix missing text when rendering Beamer PDFs that use Type3 fonts.

Some Beamer PDFs render their glyphs as inline image masks inside Type3
`CharProcs`. Previously the CPU renderer ignored those `DrawImage` commands,
and Type3 glyph advances did not apply the font's `FontMatrix`. As a result,
the page could contain backgrounds and images but no visible text, or have
incorrect text spacing.

This change:

- applies `FontMatrix` when calculating Type3 glyph advances;
- renders image commands inside Type3 glyph streams;
- recolors image masks using the current text paint;
- avoids placing temporary Type3 glyph images into the page-level downscaled
  image cache;
- adds a regression test for Type3 glyph advance calculation.

Example PDF:
[CTAN PDF：beamerexample-conference-talk.pdf](https://tug.ctan.org/macros/latex/contrib/beamer/doc/beamerexample-conference-talk.pdf)
## Related issue

Related: none

## Implementation notes

Type3 glyph widths are stored in glyph space and must be transformed through
the font's `/FontMatrix` before being used for text positioning. This follows
the Type3 font model described in ISO 32000, section 9.6.6.

Beamer's Computer Modern Type3 fonts may define glyph shapes with
`/ImageMask` inline images instead of vector paths. The nested content
interpreter now collects those images in a local `ImageCache`. Since image
masks are stencils rather than ordinary black images, their opaque pixels are
tinted with the current text paint before rendering.

The patch does not change public APIs, add dependencies, or change the GPU
renderer. It only extends the existing CPU Type3 rendering path.

No PDF fixture was added because the original failing PDF is not part of the
zpdf repository. The behavior was verified with a real Beamer PDF locally.

## Validation

```text
cargo fmt --all --check
# passed
cargo check -p zpdf-font -p zpdf-render-cpu

# passed
cargo test -p zpdf-font -p zpdf-render-cpu
# passed:
# - zpdf-font: 77 unit tests
# - variable_font: 4 tests
# - zpdf-render-cpu: 43 unit tests

cargo run -p zpdf-cli -- render beamerexample-conference-talk.pdf \
  -p 1 -o /tmp/zpdf-beamer-type3-page1.png --dpi 72 --backend cpu
# passed; Type3 text was visible in the rendered output

cargo clippy --workspace --all-targets -- -D warnings

cargo test --workspace
```

The full workspace Clippy and test commands were not run locally. The focused
checks above passed.

## Compatibility and risk

- No public API changes.
- No new dependencies.
- No changes to the GPU renderer.
- CPU rendering of Type3 fonts using vector CharProcs remains unchanged.
- Type3 fonts using image masks now render their glyphs and text colors
correctly.
- Temporary glyph images are rendered without entering the persistent
downscaled image cache.
## Checklist

- [X] The change is focused and does not include unrelated refactoring.
- [X] I added or updated tests, or explained above why tests are not practical.
- [X] I ran the relevant formatting, lint, build, and test commands.
- [ ] I updated public documentation and `docs/CHANGELOG.md` for user-visible changes.
- [X] New or changed fixtures are minimal, redistributable, and contain no confidential data.
- [X] I called out breaking API or behavior changes explicitly.
- [X] If AI tools were used, I followed [AI_POLICY.md](../../AI_POLICY.md) and added a short human-written note in my own words (my mother tongue) at the bottom of this PR explaining what I intended to do.

## Human note

我想修复一份 Beamer PDF 中 Type3 字体无法显示的问题, 这导致整个 PDF 中所有的文字都无法显示. 我调查后发现是字体宽度需要使用 FontMatrix, 并且 CharProc 中的 ImageMask 也需要安装当前文字的颜色绘制, 我检查了这次修改, 并且用实际的 Beramer PDF 验证了结果. 在代码分析和修复过程中, 我使用了 AI 辅助, 但是最终的改动和结果验证都经过了我自己的人工确认.

### CODE REVIEW

**Comment by YUZHEthefool** [comment]
Changes makes sense. Some suggestions. Thanks

**Review by YUZHEthefool — COMMENTED** [review]
I verified that the supplied Beamer PDF renders text with this PR, while the base revision does not. Before merge, please address the four edge cases below; I reproduced each with a small synthetic display list, and the current tests do not cover them.

File: crates/zpdf-render-cpu/src/lib.rs [hunk 1]
```diff
@@ -1946,7 +1968,10 @@ impl<'a> CpuRenderer<'a> {
             };
 
             let glyph_rect = zpdf_core::Rect::new(0.0, -1000.0, 1000.0, 1000.0);
-            let glyph_dl = ContentInterpreter::new(glyph_rect).interpret(stream);
+            let mut glyph_images = ImageCache::new();
+            let glyph_dl = ContentInterpreter::new(glyph_rect)
```
Comment (YUZHEthefool) [inline crates/zpdf-render-cpu/src/lib.rs:1972]: This nested interpreter is created with `ParseLimits::default()`. Once CharProcs can decode images, this path bypasses the document's `max_image_pixels` and `max_image_cache_bytes` limits during rendering. I reproduced this with `max_image_pixels = 1`: an 8x8 inline mask still renders. Please retain the document limits in `CpuRenderer` (or pass them into this method), call `.with_limits(...)` here, and add a regression test for the limit.

Concretely (`crates/zpdf-render-cpu/src/lib.rs`):

```rust
// CpuRenderer: keep the document's limits, as `with_limits` already does for
// the render budgets.
parse_limits: ParseLimits,

// with_limits()
self.parse_limits = limits.clone();

// render_type3_glyphs()
let glyph_dl = ContentInterpreter::new(glyph_rect)
    .with_images(&mut glyph_images)
    .with_limits(&self.parse_limits)
    .interpret(stream);
```

File: crates/zpdf-render-cpu/src/lib.rs [hunk 2]
```diff
@@ -2008,12 +2033,85 @@ impl<'a> CpuRenderer<'a> {
...
+        let Some(tint) = (match &paint.shader {
+            tiny_skia::Shader::SolidColor(color) => Some(*color),
+            _ => None,
+        }) else {
+            return;
+        };
+        let Some(source) = image_cache.get(draw.image_id) else {
+            return;
+        };
+        if !source.has_alpha {
```
Comment (YUZHEthefool) [inline crates/zpdf-render-cpu/src/lib.rs:2085]: `has_alpha` is not an `/ImageMask` discriminator: it is also set for color-key `/Mask`, `/SMask`, and ordinary RGBA/JPX images. This branch recolors every such image to the outer text paint, so a normal transparent image in a CharProc loses its source RGB. Please carry an explicit ImageMask bit through `DecodedImage`/`ImageDraw` and tint only that case; add a regression for a non-mask transparent image.

Sketch:

```rust
// zpdf-image: set this only for /ImageMask (and inline /IM true).
pub struct DecodedImage {
    // ...
    pub is_image_mask: bool,
}

// zpdf-display-list: carry it to the backend.
pub struct ImageDraw {
    // ...
    pub is_image_mask: bool,
}

// here: tint stencils only.
if !draw.is_image_mask {
    self.render_image_from_cache(image_cache, draw, alpha, false);
    return;
}
```

File: crates/zpdf-render-cpu/src/lib.rs [hunk 3]
```diff
@@ -2008,12 +2033,85 @@ impl<'a> CpuRenderer<'a> {
...
+                                font_matrix[4],
+                                font_matrix[5],
+                            ))
+                            .concat(&draw.transform);
+                        let draw = ImageDraw {
+                            image_id: draw.image_id,
+                            transform: glyph_to_page,
+                            alpha: draw.alpha,
+                        };
+                        self.render_type3_image(&glyph_images, &draw, paint, run.alpha);
```
Comment (YUZHEthefool) [inline crates/zpdf-render-cpu/src/lib.rs:2057]: `render_glyph_run_with_alpha` deliberately accepts an `alpha_override` for the knockout shape pass, but this call hard-codes `run.alpha`. In `render_shape_cmd` the override is `1.0`, so Type3 image masks capture only the original alpha and knockout compositing differs from an equivalent vector glyph. Please thread `alpha_override` through `render_type3_glyphs` and `render_type3_image`, combine it with `draw.alpha`, and add a transparent-mask knockout regression.

Sketch:

```rust
// render_glyph_run_with_alpha
self.render_type3_glyphs(run, font, &paint, alpha_override);

// render_type3_glyphs(..., alpha_override: f32)
let alpha = alpha_override * draw.alpha;
self.render_type3_image(&glyph_images, &draw, paint, alpha);
```

File: crates/zpdf-font/src/lib.rs [hunk 4]
```diff
@@ -825,6 +825,14 @@ impl LoadedFont {
     /// [`advance_divisor`](Self::advance_divisor). Prefers the PDF /Widths entry
     /// (keyed by character *code*, authoritative per the spec) over the font's hmtx.
     pub fn simple_glyph_advance(&self, code: u16, gid: u16) -> f64 {
+        if let PdfFontType::Type3 { font_matrix, .. } = &self.font_type {
+            // Type 3 /Widths are expressed in glyph space, so apply the
+            // font matrix before converting them to the renderer's normalized
+            // font-unit scale. Simple-font /Widths are already in 1/1000
+            // text-space units and must not take this path.
+            let scale = font_matrix[0].hypot(font_matrix[1]).max(1e-12);
```
Comment (YUZHEthefool) [inline crates/zpdf-font/src/lib.rs:833]: `hypot(font_matrix[0], font_matrix[1])` is the length of the transformed glyph x-axis, not the horizontal Type3 advance. A shear or rotation therefore changes text spacing (`[0.001, 0.001, 0, 0.001, ...]` turns a 1000-unit width into 1414). Please transform the PDF width vector consistently with the renderer's text-matrix representation instead of folding `b` into a scalar horizontal advance, and add a non-diagonal FontMatrix regression test.

For the current scalar advance representation, the minimal local correction is:

```suggestion
let scale = font_matrix[0];
```

**Comment by YUZHEthefool** [comment]
If you have any further questions for this pr, feel free to @ me, and I’ll reply as soon as possible :)

**Comment by imkerberos** [comment]
Thanks for the detailed review. I addressed all four edge cases in 879fc71:

- nested Type3 interpretation now inherits the document ParseLimits;
- ImageMask is distinguished from ordinary transparent images;
- Type3 image alpha is combined with the outer alpha override;
- Type3 glyph advance uses the horizontal FontMatrix component.

I also added regression tests for these cases. The full workspace test suite passes.
Could you please take another look?

**Comment by YUZHEthefool** [comment]
@xero-team-bot r+

**Comment by YUZHEthefool** [comment]
thanks!
